// react-native-element/Library/react-native-element-datepickview/EDatePickerView.ios.js
/**
 * Created on 12:30 2018/08/14.
 * file name Picker
 * by wangtieshan
 */

import React, {Component} from 'react';

import {
    View,
    Modal,
    TouchableOpacity,
    Text,
    StyleSheet,
    DatePickerIOS,
} from 'react-native';

import PropTypes from 'prop-types';

export interface EDatePickerViewOptions {
    /**
     * 选中时间
     */
    date: any;

    /**
     * 提示语
     */
    prompt: string;

    /**
     * 选中回调
     */
    onSelectedDate: (date: string)=>void;

    /**
     * Maximum date.
     */
    maximumDate: any;

    /**
     * Minimum date.
     */
    minimumDate: any;

    /**
     * The date picker mode.['date', 'time', 'datetime']
     */
    mode: string;

    /**
     * The interval at which minutes can be selected.
     * [1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30]
     */
    minuteInterval: number;
}

Object.assign(Component.prototype, {

    datePickerView() {
        if (!this.state) {
            this.state = {};
        }
        if (!this.state.datePicker) {
            this.state.datePicker = {visible: false};
        }
        return (
            <EDatePickerView
                {...this.state.datePicker}
                onDateChange={this.state.datePicker.onDateChange}
                date={this.state.datePicker.date}
                onNeedDismiss={()=>{
                    let sheet = this.state.datePicker;
                    sheet.visible = false;
                    this.setState({
                        datePicker: sheet
                    });
                }}/>
        )
    },

    /**
     * alert
     * @param options EDatePickerViewOptions
     */
    pickDate(options: EDatePickerViewOptions) {
        this.setState({
            datePicker: {
                ...options,
                visible: true,
            }
        });
    }
});

export default class EDatePickerView extends React.Component {

    static defaultProps = {
        visible: false,
        onSelectedValue: ()=>{},
        prompt: '请选择',
        date: new Date(),
        onDateChange: ()=>{},
    };

    static propTypes = {
        ...DatePickerIOS.propTypes,
        /*是否展示*/
        visible: PropTypes.bool.isRequired,
        /*选中回调*/
        onSelectedValue: PropTypes.func,
        /*当需要消失时的回调*/
        onNeedDismiss: PropTypes.func.isRequired,

        /*提示语*/
        prompt: PropTypes.string,
    };

    /*临时存储数据*/
    tmpSelected: string;

    constructor(props) {
        super(props);
        this.state = {
            selected: props.date,
        };
    }

    componentWillReceiveProps(props) {
        this.setState({
            selected: props.date,
        })
    }

    _onValueChanged(selected) {
        this.setState({
            selected: selected,
        });
        if (this.props.onDateChange) {
            this.props.onDateChange(selected);
        }
    }

    _onEnsure() {
        if (!this.state.selected || !this.state.selected.getFullYear()) {
            return;
        }
        this.tmpSelected = this.state.selected;
        this._onNeedDismiss();
    }

    _onDismissed() {
        if (this.tmpSelected) {
            this.props.onSelectedValue(this.tmpSelected);
            this.tmpSelected = null;
        }
    }

    _onCancel() {
        this.state.selected = null;
        this._onNeedDismiss();
    }

    _onNeedDismiss() {
        this.props.onNeedDismiss()
    }

    render() {
        return (
            <Modal transparent={true}
                   animationType={'slide'}
                   visible={this.props.visible}
                   onDismiss={()=>{this._onDismissed()}}
                   onRequestClose={()=>this._onNeedDismiss()}>
                <TouchableOpacity style={styles.body}
                                  activeOpacity={1}
                                  onPress={()=>this._onNeedDismiss()}>
                    <TouchableOpacity style={styles.barStyle} activeOpacity={1}>
                        <TouchableOpacity style={{justifyContent: 'center',}} onPress={()=>this._onCancel()}>
                            <Text style={{color: 'black', fontSize: 16, marginLeft: 20,}}>取消</Text>
                        </TouchableOpacity>
                        <View style={{flex: 1, justifyContent: 'center', alignItems: 'stretch',}}>
                            <Text style={{color: 'gray', fontSize: 16, marginRight: 20, textAlign: 'center',}}>{this.props.prompt}</Text>
                        </View>
                        <TouchableOpacity style={{justifyContent: 'center',}} onPress={()=>{this._onEnsure()}}>
                            <Text style={{color: 'black', fontSize: 16, marginRight: 20,}}>确定</Text>
                        </TouchableOpacity>
                    </TouchableOpacity>
                    <DatePickerIOS {...this.props} style={{backgroundColor: '#f8f8f8',}} date={this.state.selected} onDateChange={(date)=>{this._onValueChanged(date)}}/>
                </TouchableOpacity>
            </Modal>
        )
    }

    _getItems() {
        let result = [];
        for (let item of this.props.items) {
            result.push(<Picker.Item label={item} value={item} key={item}/>)
        }
        return result;
    }
}

const styles = StyleSheet.create({
    body: {
        flex: 1,
        backgroundColor: 'transparent',
        justifyContent: 'flex-end',
        alignItems: 'stretch',
        shadowOffset: {height: -3,},
        shadowColor: 'gray',
        shadowOpacity: 0.1,
    },
    barStyle: {
        height: 40,
        flexDirection: 'row',
        backgroundColor: 'white',
        shadowOffset: {height: 3,},
        shadowColor: '#eee',
        shadowOpacity: 0.5,
        zIndex: 100,
    },
});
