// react-native-element/Library/react-native-element-list/EListItem.js
/**
 * Created on 16:48 2019/03/12.
 * file name EListItem
 * by wangtieshan
 */

import React from 'react';

import {
    EButton
} from '../se';

import {
    View,
    ColorPropType
} from 'react-native';

import PropTypes from 'prop-types';

import { ETheme } from "../react-native-element-core/theme";

export class EListItem extends React.Component {

    static propTypes = {
        ...EButton.propTypes,
        /**
         * content's css。the super view's css of title or the super view's css of `customContent`
         */
        containerStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
        /**
         * separator
         */
        separatorStyle: PropTypes.shape({
            left: PropTypes.number,
            right: PropTypes.number,
            height: PropTypes.number,
            backgroundColor: ColorPropType
        })
    };

    _separatorStyle = {
        position: 'absolute',
        bottom: 0,
        right: 0,
        height: 0.4,
        backgroundColor: ETheme.color.separatorColor
    };

    _buttonContainerStyle = {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'flex-start'
    };

    render() {
        return (
            <EButton {...this.props}
                     borderConfig={{ borderRadius: 0 }}
                     containerStyle={[this._buttonContainerStyle, this.props.containerStyle]}
                     style={[{ backgroundColor: 'white', height: null }, this.props.style]}>
                {this.props.children}
                <View
                    disabled={true}
                    style={[this._separatorStyle, { left: ETheme.css.row.sideDistance }, this.props.separatorStyle]}/>
            </EButton>
        )
    }
}
