// react-native-element/Library/react-native-element-list/EScrollView.js

import React, {Component} from 'react';

import PropTypes from 'prop-types';

import {
    ScrollView,
    RefreshControl,
} from 'react-native';

/**
 * 增加下拉刷新的功能
 *
 * eg:
 *
 * onRefresh(end) {
 *  // ... net work or long time
 *
 *  // stop refresh
 *  end()
 * }
 *
 * <EScrollView onRefresh={end=>this.onRefresh(end)}></EScrollView>
 */
export class EScrollView extends Component {

    static propTypes = {
        /**
         * 下拉刷新回调
         * onRefresh: (end)=>void
         * 调用 end 可以停止下拉刷新
         */
        onRefresh: PropTypes.func,
    };

    constructor(props) {
        super(props);
        this.state = {
            isRefreshing: false
        };
    }

    _onRefresh() {
        this.setState({
            isRefreshing: true,
            title: '加载中...',
        });
        if (this.props.onRefresh) {
            this.props.onRefresh(()=>{
                this.setState({
                    isRefreshing: false,
                    title: '下拉加载...',
                });
            })
        }
    }

    render() {
        let refresh = {};
        if (this.props.onRefresh) {
            refresh.refreshControl = <RefreshControl
                refreshing={this.state.isRefreshing}
                onRefresh={this._onRefresh.bind(this)}
                tintColor={"gray"}
                title={this.state.title}
                titleColor={"#4a4a4a"}
                colors={['#000000']}
                progressBackgroundColor="#fff"
            />
        }
        return (
            <ScrollView {...this.props}
                        {...refresh}
                        style={[{ flex: 1, alignSelf: 'stretch', }, this.props.style]}
            />
        )
    }

}
