// react-native-element/Library/react-native-element-check/ECheck.js
/**
 * Created by wangtieshan on 2017/10/30.
 */

/* 检查某个对象是否存在指定的key*/
export function checkObjExistKey(obj: any, key: string) {
    return obj.hasOwnProperty(key) && !!obj[key] && (obj[key] !== '');
}

/*删除空格*/
export function removeSpace(c: string): string {
    if (!c) {return c}
    return c.replace(/ /g,'')
}

/**
 * 姓名中间点转化
 * @param c
 */
export function replacePoint(c: string) {
    if (!c) {return c;}
    return c.replace(/[.,．，。]/g, '·');
}

/*空检查*/
export function checkEmpty(num: string): boolean {
    if (!num) {
        return false;
    }
    if (typeof(num) !== 'string') {return false}
    let reg = /^.+$/g;
    return reg.test(num)
}

/*纯数字检查*/
export function checkOnlyNum(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg = /^\d+$/g;
    return reg.test(num)
}

/*身份证检测*/
export function checkIdNum(num: string): string {
    var reg = /^(\d{14}|\d{17})(\d|[xX])$/g;
    if (!reg.test(num)) {
        return false;
    }
    var dataStr = num.substr(6,8);
    var year = dataStr.substr(0,4);
    var month = dataStr.substr(4,2);
    var day = dataStr.substr(6,2);
    var date = year + '-' + month + '-' + day;
    return (new Date(date).getDate()==date.substring(date.length-2));
}

/*姓名检查*/
export function checkPersonName(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg =new RegExp("^([\\\u4e00-\\\u9fa5]|[·]){1,20}$","g");
    return reg.test(num)
}

/*银行卡校验*/
export function checkBankNum(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg = /^[0-9]{16,19}$/g;
    return reg.test(num)
}

/*数字或大小写字母*/
export function checkNumOrCharacter(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg = /^[0-9A-Za-z]+$/g;
    return reg.test(num)
}

/*数字或大写字母*/
export function checkNumOrUpperCharacter(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg = /^[0-9A-Z]+$/g;
    return reg.test(num)
}

/*数字或小写字母*/
export function checkNumOrLCharacter(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg = /^[0-9a-z]+$/g;
    return reg.test(num)
}

/*电话号码，手机号加座机号*/
export function checkPhone(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    return checkMobilePhone(num) || checkLandline(num)
}

/*手机号校验*/
export function checkMobilePhone(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg = /^1[0-9]{10}$/g;
    return reg.test(num)
}

/*座机校验*/
export function checkLandLine(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg = /^[0-9]+-[0-9]+$/g;
    return reg.test(num)
}

/*金额校验*/
export function checkMoney(num: string): boolean {
    if (!checkEmpty(num)) {return false}
    let reg = /^[0-9]+\.[0-9]+$/g;
    return reg.test(num)
}

export const ECheck = {
    checkLandLine,
    checkMoney,
    checkMobilePhone,
    removeSpace,
    replacePoint,
    checkEmpty,
    checkOnlyNum,
    checkIdNum,
    checkPersonName,
    checkBankNum,
    checkNumOrCharacter,
    checkNumOrUpperCharacter,
    checkNumOrLCharacter,
    checkPhone,
    checkObjExistKey
};
