// react-native-element/Library/react-native-element-row/ERowSegmentControl.js
/**
 * Created on 14:06 2019/03/13.
 * file name ERowSegmentControl
 * by wangtieshan
 */

import React from 'react';

import { ESegmentControl } from '../react-native-element-core/ESegmentControl';

import {
    ETheme
} from '../react-native-element-core/theme';

import {
    View,
    Text
} from 'react-native';

import PropTypes from 'prop-types';

export class ERowSegmentControl extends React.Component{

    static propTypes = {

        title: PropTypes.string,

        titleMinWidth: PropTypes.number,

        titleStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),

        segmentControlContainerStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    };

    static defaultProps = {

    };

    _segmentControlContainerStyle = {
        flex: 1,
        alignItems: 'flex-start'
    };

    render() {
        return (
            <View style={[{
                paddingHorizontal: ETheme.css.row.sideDistance,
                flexDirection: 'row',
                height: 44,
                alignItems: 'center',
                backgroundColor: 'white' }, this.props.style]}>
                <Text style={[{minWidth: this.props.titleMinWidth}, this.props.titleStyle]}>
                    {this.props.title}
                    </Text>
                <View style={[this._segmentControlContainerStyle, this.props.segmentControlContainerStyle]}>
                    <ESegmentControl {...this.props} style={[this.props.segmentControlStyle]}/>
                </View>
            </View>
        )
    }

}
