// react-native-element/Library/react-native-element-core/ENavigationBackButton.js
/**
 * Created on 09:47 2019/03/26.
 * file name ENavigationBackButton
 * by wangtieshan
 */

import React from "react";

import {
    EButtonOpacity
} from "./EButton";

import { ColorPropType, Text } from 'react-native';

import { EImage } from './EImage';

import PropTypes from 'prop-types';

/**
 * 导航返回按钮
 */
export class ENavigationBackButton extends React.Component {

    static propTypes = {
        /**
         * 标题
         */
        title: PropTypes.string,
        /**
         * 颜色
         */
        tintColor: ColorPropType,
        /**
         * 点击回调
         */
        onPress: PropTypes.func,
        /**
         * 按下去之后更改的透明度
         */
        activeOpacity: PropTypes.number
    };

    static defaultProps = {
        tintColor: '#000000',
        activeOpacity: 0.4,
        showBackImage: true
    };

    _buttonStyle = {
        paddingLeft: 10,
        height: 44,
        minWidth: 80
    };

    _containerStyle = {
        flexDirection: 'row',
        justifyContent: 'flex-start'
    };

    _imageStyle = {
        width: 11,
        height: 20
    };

    render() {
        return (
            <EButtonOpacity
                activeOpacity={this.props.activeOpacity}
                onPress={()=>this.props.onPress && this.props.onPress()}
                style={this._buttonStyle}
                containerStyle={this._containerStyle}>
                { this.props.showBackImage && <EImage
                    contentMode={'contain'}
                    tintColor={this.props.tintColor}
                    source={require('./ArrowLeft.png')}
                    style={[this._imageStyle]}/> }
                <Text
                    style={{ fontSize: 16, marginLeft: 5, color: this.props.tintColor }}>
                    { this.props.title || '' }
                    </Text>
            </EButtonOpacity>
        )
    }
}

/**
 * 导航按钮
 * 一般用来设置右上角的按钮
 */
export class ENavigationButtonItem extends ENavigationBackButton {

    static defaultProps = {
        ...ENavigationBackButton.defaultProps,
        showBackImage: false
    };

    _buttonStyle = {
        height: 44,
        paddingHorizontal: 10,
        minWidth: 35
    };

}
