// react-native-element/Library/react-native-element-core/ESep.js
/**
 * Created on 11:03 2019/04/01.
 * file name ESep
 * by wangtieshan
 */

import React from 'react';

import {
    View,
    ColorPropType
} from 'react-native';

import {
    ETheme
} from './theme';

import PropTypes from 'prop-types';

/**
 * 分割线组件
 */
export class ESep extends React.Component {

    static propTypes = {
        /**
         * 分割线颜色
         */
        color: ColorPropType,
        /**
         * 是否采用绝对布局
         * 如果为绝对布局，则绝对布局在底部
         * 如果为false，则为相对布局
         */
        absolute: PropTypes.bool,
        /**
         * 分割线高度
         */
        height: PropTypes.number,
        /**
         * 左侧缩进
         */
        left: PropTypes.number,
        /**
         * 右侧缩进
         */
        right: PropTypes.number,
        /**
         * 底部缩进
         */
        bottom: PropTypes.number,
    };

    static defaultProps = {
        color: ETheme.color.separatorColor,
        height: 0.4,
        left: ETheme.css.row.sideDistance,
        right: 0,
        bottom: 0,
        absolute: true
    };

    render() {
        return <View style={this._style()}/>;
    }

    _style() {
        let d;
        if (this.props.absolute) {
            d = {
                position: 'absolute',
                left: this.props.left,
                right: this.props.right,
                bottom: this.props.bottom
            };
        } else {
            d = {
                position: 'relative',
                marginLeft: this.props.left,
                marginRight: this.props.right
            };
        }
        return [d, { backgroundColor: this.props.color, height: this.props.height }, this.props.style];
    }
}
