// react-native-element/Library/react-native-element-core/theme.js
/**
 * Created on 14:37 2019/03/06.
 * file name theme
 * by wangtieshan
 */

export interface EThemeInterface {

    /* 颜色 */
    color: EColorInterface;

    /* css */
    css: ECssInterface;

    /* 字体 */
    font: EFontInterface;
}

export interface EColorInterface {

    /**
     * 主色调
     */
    main: string;

    /**
     * 与主色调匹配的文本颜色。
     */
    textColorMatchMain: string;

    /**
     * 页面背景色
     */
    pageBackgroundColor: string;

    /**
     * modal视图的背景颜色
     */
    modalBackgroundColor: string;

    /**
     * 各种状态不可用下的颜色，一般用来做背景色，例如：按钮不可用下的背景色
     */
    disabledColor: string;

    /**
     * 各种不可用状态下的文字的颜色。例如：按钮不可用下文本颜色
     */
    disabledTextColor: string;

    /**
     * 占位文本颜色
     */
    placeholderTextColor: string;

    /**
     * 分割线颜色
     */
    separatorColor: string;

    color4a: string;

    color9b: string;
}

export interface EFontInterface {

    /**
     * 字体大小
     */
    size: EFontSizeInterface;

    /**
     * 行高
     */
    lineHeight: EFontLineHeightInterface;
}

export interface EFontLineHeightInterface {

    /**
     * 无行高 1.0
     */
    none: number;
    /**
     * 紧凑 1.3
     */
    compact: number;
    /**
     * 常规 1.5
     */
    regular: number;
    /**
     * 宽松 1.7
     */
    loose: number;
}

export interface EFontSizeInterface {
    /**
     * 最小字体 10px mini
     */
    mini: string;

    /**
     * 辅助文字	12px Extra Small
     */
    extraSmall: string;

    /**
     * 正文（小）	13px Small
     */
    small: string;

    /**
     * 正文	14px Base
     */
    base: string;

    /**
     * 小标题	16px Medium
     */
    medium: string;

    /**
     * 标题	18px large
     */
    large: string;

    /**
     * 主标题	20px Extra large
     */
    extraLarge: string;
}

export interface ECssInterface {

    /**
     * 横向列表布局控件配置
     */
    row: ECssRow;
}

/**
 * 行控件配置
 */
export interface ECssRowInterface {
    /**
     * 到边缘的距离
     */
    sideDistance: number;
}

export let ETheme: EThemeInterface = {
    color: {
        main: '#ffe600',
        textColorMatchMain: 'black',
        disabledColor: '#EAEAEA',
        disabledTextColor: '#9B9B9B',
        modalBackgroundColor: '#00000099',
        color4a: '#4a4a4a',
        color9b: '#9b9b9b',
        placeholderTextColor: '#c6c6c6',
        separatorColor: '#a7a7aa',
        pageBackgroundColor: '#f8f8f8'
    },
    css: {
        row: {
            sideDistance: 15
        }
    },
    font: {
        size: {
            mini: 10,
            extraSmall: 12,
            small: 13,
            base: 14,
            medium: 16,
            large: 18,
            extraLarge: 20
        },
        lineHeight: {
            none: 1.0,
            compact: 1.3,
            regular: 1.5,
            loose: 1.7
        }
    }
};
