// react-native-element/Library/react-native-element-core/API/ENativeSystemApi.js
/**
 * Created on 22:58 2018/06/14.
 * file name ENativeSystemApi
 * by wangtieshan
 */

import {
    NativeModules,
} from 'react-native';

export const EMiniNativeSystemAPI = NativeModules.EMiniSystemAPI;

/**
 * 通讯录选择回调
 */
export interface EContactType {
    name: string;
    phone: string;
}
/**
 * 定位信息
 */
export interface ELocationType {
    latitude: string;
    longitude: string;
}
/**
 * 获取定位信息
 * @param callBack
 */
export function getLocation(callBack: (ELocationType)=>void) {
    EMiniNativeSystemAPI.getLocation((error, result)=>{
        if (result) {
            callBack(JSON.parse(result))
        }
    })
}

/**
 * 通讯录选择
 * @param callBack
 */
export function contactPick(callBack: (EContactType)=>void) {
    EMiniNativeSystemAPI.contactPick((error, result)=>{
        if (result) {
            callBack(JSON.parse(result))
        } else {
            callBack({});
        }
    })
}

/**
 * 获取所有通讯录联系人
 */
export function getAllContacts(callBack: EContactType[]) {
    EMiniNativeSystemAPI.getAllContacts((error, data: EContactType[])=>{
        callBack(data);
    });
}

/**
 * 拨打手机号
 * @param phoneNum
 */
export function phoneCall(phoneNum: string) {
    EMiniNativeSystemAPI.phoneCall(phoneNum);
}

/**
 * 手机震动
 * @param type light medium heavy auto
 */
export function vibrate(type: string) {
    if (EMiniNativeSystemAPI.vibrate) {
        EMiniNativeSystemAPI.vibrate(type)
    }
}

/**
 * 系统API
 */
export interface ENativeSystemAPIInterface {

    /**
     * 获取定位信息
     */
    getLocation: ((ELocationType)=>void) => void;

    /**
     * 通讯录选择
     */
    contactPick: ((EContactType)=>void)=>void;

    /**
     * 获取系统通讯录所有人员
     */
    getAllContacts: ((EContactType[]) => void) => void;

    /**
     * 调起震动
     */
    vibrate: (string) => void;

    /**
     * 打电话
     */
    phoneCall: (string) => void;
}
