// react-native-element/Library/react-native-element-imagetool/EImageTool.js
/**
 * Created by wangtieshan on 2017/10/29.
 */

import React from 'react';

import {
    NativeModules
} from 'react-native';

const EImageNativeTool = NativeModules.EImageTool;

/**
 * 图片公共属性
 */
export interface EImageProperty {
    path: string,
    // 相对路径可能是不存在的
    subPath: string,
    width: number,
    height: number,
    scale: number,
}

/**
 * 图片压缩配置
 * maxWidth: 压缩后的最大宽度
 * maxHeight: 压缩后的最大高度
 * size: 压缩后的最大大小单位：M
 */
export interface EImageCompressConfig {
    maxWidth: number,
    maxHeight: number,
    size: number,
}

/**
 * 选择图片的配置项
 * sourceType?: album or camera
 * cameraDevice?: front back
 */
export type EPickImageConfig = {

    sourceType?: string;

    cameraDevice?: string;
}

export class EImageTool {

    /**
     * 保存图片到相册
     * @param path
     * @param callBack
     */
    static saveToAlbum(path, callBack: {error:{code: string, message: string}}) {
        EImageNativeTool.saveToAlbum(path, callBack);
    }

    /**
     * 图片压缩
     * @param path 当前图片的录几个
     * @param config 压缩配置
     * @param callback 结果 EImageProperty
     */
    static compressImage(path, config: EImageCompressConfig, callback: (EImageProperty) => void) {
        EImageNativeTool.compressImage(path, JSON.stringify(config || { maxWidth: 1024, maxHeight: 1024, size: 0.7}), (error, data)=>{
            if (callback) {
                if (data) {
                    callback(JSON.parse(data));
                } else {
                    callback(null);
                }
            }
        })
    }

    /**t
     * 从相册选一张照片
     * @param options 选择图片配置
     */
    static pickImage(options: { config: EPickImageConfig, res: (EImageProperty) => void }) {
        if (!options.res) { return }
        let config = options.config || {};
        config.cameraDevice = config.cameraDevice || 'back';
        config.sourceType = config.sourceType || 'camera';
        EImageNativeTool.pickImageFromAlbum(JSON.stringify(config), (error, data)=>{
            if (data) {
                options.res(JSON.parse(data));
            } else {
                options.res(null);
            }
        })
    }
}
