// react-native-element/Library/react-native-element-mini/EMiniLifeCycle.js
/**
 * Created on 13:56 2018/05/13.
 * file name EMiniLifeCycle
 * by wangtieshan
 */

import { EMiniEvent } from './EMiniEvent';

const _eventSubscriptions = {
    viewDidAppear: new Set(),
    viewWillAppear: new Set(),
    viewDidDisappear: new Set(),
    viewWillDisappear: new Set(),
};

/**
 * 小程序原生页面声明周期
 * @type {{_init: ENativeLifeCycle._init, viewDidAppear: (function(function(): void): {remove: (function(): boolean)}), viewWillAppear: (function(function(): void): {remove: (function(): boolean)}), viewDidDisappear: (function(function(): void): {remove: (function(): boolean)}), viewWillDisappear: (function(function(): void): {remove: (function(): boolean)})}}
 */
const ENativeLifeCycle = {

    _init: function () {

        EMiniEvent.register('viewWillAppear', function() {
            const subscriptions = [..._eventSubscriptions.viewWillAppear].reverse();
            subscriptions.forEach(function(item) {
                item();
            });
        });

        EMiniEvent.register('viewDidAppear', function() {
            const subscriptions = [..._eventSubscriptions.viewDidAppear].reverse();
            subscriptions.forEach(function(item) {
                item();
            });
        });

        EMiniEvent.register('viewWillDisappear', function() {
            const subscriptions = [..._eventSubscriptions.viewWillDisappear].reverse();
            subscriptions.forEach(function(item) {
                item();
            });
        });

        EMiniEvent.register('viewDidDisappear', function() {
            const subscriptions = [..._eventSubscriptions.viewDidDisappear].reverse();
            subscriptions.forEach(function(item) {
                item();
            });
        });
    },

    viewDidAppear: function (
        handler: ()=>void
    ): {remove: () => void} {
        _eventSubscriptions.viewDidAppear.add(handler);
        return {
            remove: () => _eventSubscriptions.viewDidAppear.delete(handler),
        };
    },

    viewWillAppear: function (
        handler: ()=>void
    ): {remove: () => void} {
        _eventSubscriptions.viewWillAppear.add(handler);
        return {
            remove: () => _eventSubscriptions.viewWillAppear.delete(handler),
        };
    },

    viewDidDisappear: function (
        handler: ()=>void
    ): {remove: () => void} {
        _eventSubscriptions.viewDidDisappear.add(handler);
        return {
            remove: () => _eventSubscriptions.viewDidDisappear.delete(handler),
        };
    },

    viewWillDisappear: function (
        handler: ()=>void
    ): {remove: () => void} {
        _eventSubscriptions.viewWillDisappear.add(handler);
        return {
            remove: () => _eventSubscriptions.viewWillDisappear.delete(handler),
        };
    },
};

ENativeLifeCycle._init();

export { ENativeLifeCycle };
